### PR TITLE
Builtin rule to lift model methods to updates

### PR DESCRIPTION
**Pull request by mattulbrich**
This is a fix for #1465. See there for details.

### CODE REVIEW

**Comment by mattulbrich** [comment]
While this fixed the assignment of model *methods*, model *fields* still raise a differentexception if assigned to ghost fields/vars.

Let's fix this, while we are on it ...

```
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
	at de.uka.ilkd.key.java.TypeConverter.convertToLogicElement(TypeConverter.java:334)
	at de.uka.ilkd.key.java.TypeConverter.convertToLogicElement(TypeConverter.java:301)
	at de.uka.ilkd.key.java.TypeConverter.convertReferencePrefix(TypeConverter.java:180)
	at de.uka.ilkd.key.java.TypeConverter.convertToLogicElement(TypeConverter.java:319)
	at de.uka.ilkd.key.rule.inst.SVInstantiations.getTermInstantiation(SVInstantiations.java:411)
	at de.uka.ilkd.key.rule.SyntacticalReplaceVisitor.visit(SyntacticalReplaceVisitor.java:289)
	at de.uka.ilkd.key.logic.TermImpl.execPostOrder(TermImpl.java:264)
	at de.uka.ilkd.key.logic.TermImpl.execPostOrder(TermImpl.java:261)
	at de.uka.ilkd.key.logic.TermImpl.execPostOrder(TermImpl.java:261)
	at de.uka.ilkd.key.logic.TermImpl.execPostOrder(TermImpl.java:261)
	at de.uka.ilkd.key.rule.executor.javadl.TacletExecutor.syntacticalReplace(TacletExecutor.java:87)
	at de.uka.ilkd.key.rule.executor.javadl.TacletExecutor.instantiateReplacement(TacletExecutor.java:135)
	at de.uka.ilkd.key.rule.executor.javadl.TacletExecutor.instantiateSemisequent(TacletExecutor.java:168)
	at de.uka.ilkd.key.rule.executor.javadl.TacletExecutor.replaceAtPos(TacletExecutor.java:194)
	at de.uka.ilkd.key.rule.executor.javadl.SuccTacletExecutor.applyReplacewith(SuccTacletExecutor.java:34)
	at de.uka.ilkd.key.rule.executor.javadl.FindTacletExecutor.apply(FindTacletExecutor.java:104)
	at de.uka.ilkd.key.rule.Taclet.apply(Taclet.java:935)
	at de.uka.ilkd.key.rule.TacletApp.execute(TacletApp.java:349)
	at de.uka.ilkd.key.proof.Goal.apply(Goal.java:624)

**Comment by mattulbrich** [comment]
I have also implemented the generalisation for model fields. It should get some more tests, perhaps.

**Comment by WolframPfeifer** [comment]
Something does not work for model fields yet. Consider the following test class:
```java
public final class Test {

    //@ public ghost int num;

    /*@ public model_behaviour
      @ ensures \result == 42;
      @ public static model int fourtytwo() {
      @     return 42;
      @ }
      @*/

    //@ model int mfield;
    //@ represents mfield = 42;
    //@ accessible mfield: \strictly_nothing;

    /*@ public static normal_behaviour
      @ requires mfield == 42;
      @ assignable num;
      @ ensures \result == fourtytwo();
      @ ensures \result == num;
      @*/
    public /*@ helper @*/ int test(int number) {
        //@ set num = fourtytwo();
        //@ set num = mfield;
        return 42;
    }
}
```
I always encounter the following exception when applying the new rule:
```
de.uka.ilkd.key.logic.TermCreationException: Building a term failed. Normally there is an arity mismatch or one of the subterms' sorts is not compatible (e.g. like the '2' in "true & 2")
The top level operator was Test::$mfield(Sort: int); its expected arg sorts were:
1.) sort: Heap, sort hash: 285570487
2.) sort: Test, sort hash: 904043230

The subterms were:

	at de.uka.ilkd.key.logic.op.AbstractOperator.validTopLevelException(AbstractOperator.java:86)
	at de.uka.ilkd.key.logic.op.AbstractSortedOperator.validTopLevelException(AbstractSortedOperator.java:16)
	at de.uka.ilkd.key.logic.TermImpl.checked(TermImpl.java:144)
	at de.uka.ilkd.key.logic.TermFactory.doCreateTerm(TermFactory.java:129)
	at de.uka.ilkd.key.logic.TermFactory.createTerm(TermFactory.java:64)
	at de.uka.ilkd.key.logic.TermFactory.createTerm(TermFactory.java:76)
	at de.uka.ilkd.key.logic.TermFactory.createTerm(TermFactory.java:81)
	at de.uka.ilkd.key.logic.TermBuilder.func(TermBuilder.java:395)
	at de.uka.ilkd.key.rule.metaconstruct.MemberPVToField.transform(MemberPVToField.java:31)
	at de.uka.ilkd.key.rule.SyntacticalReplaceVisitor.subtreeLeft(SyntacticalReplaceVisitor.java:420)
	at de.uka.ilkd.key.logic.TermImpl.execPostOrder(TermImpl.java:265)
	at de.uka.ilkd.key.logic.TermImpl.execPostOrder(TermImpl.java:261)
	at de.uka.ilkd.key.logic.TermImpl.execPostOrder(TermImpl.java:261)
	at de.uka.ilkd.key.logic.TermImpl.execPostOrder(TermImpl.java:261)
	at de.uka.ilkd.key.rule.executor.javadl.TacletExecutor.syntacticalReplace(TacletExecutor.java:87)
	at de.uka.ilkd.key.rule.executor.javadl.TacletExecutor.instantiateReplacement(TacletExecutor.java:135)
	at de.uka.ilkd.key.rule.executor.javadl.TacletExecutor.instantiateSemisequent(TacletExecutor.java:168)
	at de.uka.ilkd.key.rule.executor.javadl.TacletExecutor.replaceAtPos(TacletExecutor.java:194)
	at de.uka.ilkd.key.rule.executor.javadl.SuccTacletExecutor.applyReplacewith(SuccTacletExecutor.java:34)
	at de.uka.ilkd.key.rule.executor.javadl.FindTacletExecutor.apply(FindTacletExecutor.java:104)
	at de.uka.ilkd.key.rule.Taclet.apply(Taclet.java:935)
	at de.uka.ilkd.key.rule.TacletApp.execute(TacletApp.java:349)
	at de.uka.ilkd.key.proof.Goal.apply(Goal.java:624)
	at de.uka.ilkd.key.prover.impl.ApplyStrategy.applyAutomaticRule(ApplyStrategy.java:117)
	at de.uka.ilkd.key.prover.impl.ApplyStrategy.doWork(ApplyStrategy.java:144)
	at de.uka.ilkd.key.prover.impl.ApplyStrategy.executeStrategy(ApplyStrategy.java:293)
	at de.uka.ilkd.key.prover.impl.ApplyStrategy.start(ApplyStrategy.java:261)
	at de.uka.ilkd.key.ui.MediatorProofControl$AutoModeWorker.doInBackground(MediatorProofControl.java:209)
	at de.uka.ilkd.key.ui.MediatorProofControl$AutoModeWorker.doInBackground(MediatorProofControl.java:145)
	at java.desktop/javax.swing.SwingWorker$1.call(SwingWorker.java:304)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.desktop/javax.swing.SwingWorker.run(SwingWorker.java:343)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

File: key.core/src/main/java/de/uka/ilkd/key/rule/ObserverToUpdateRule.java [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+        Union<Instantiation, ModelFieldInstantiation> inst =
+            instantiate(pio.subTerm(), goal.proof().getServices());
+        if (inst == null) {
+            return false;
+        }
+
+        if (inst.isFirst()) {
+            // additional checks for method calls.
+            // currently only applicable to strictly pure methods
+            // TODO check that this is not a two-state model method.
```
Comment (WolframPfeifer) [inline key.core/src/main/java/de/uka/ilkd/key/rule/ObserverToUpdateRule.java:138]: What about this TODO? I could not figure out how to get this information here ...

**Comment by mattulbrich** [comment]
> Something does not work for model fields yet.

I added a varcond that prevents the normal field mechanisms from kicking in for model fields. The example works now.

File: key.core/src/main/java/de/uka/ilkd/key/rule/ObserverToUpdateRule.java (hunk 1 above)
Comment (mattulbrich) [reply key.core/src/main/java/de/uka/ilkd/key/rule/ObserverToUpdateRule.java]: Was not so difficult to fix. Handled in 447578efb813ba350e7da92b93c10c3eae14b034.

**Comment by WolframPfeifer** [comment]
Many tests fail here. Should we postpone?

**Review by WolframPfeifer — APPROVED** [review]
Tests are green now. Ready to merge.

Since this is the last feature for the 2.12 release, I will create the release branch as soon as this feature is merged.

**Pull request merged**